connectors/agentic-ai/src/main/java/io/camunda/connector/agenticai/aiagent/agent/AgentInitializerImpl.java

```diff
@@ -21,49 +20,64 @@
 import io.camunda.connector.agenticai.model.tool.ToolCallResult;
 import java.util.List;
 import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.springframework.util.CollectionUtils;
 
 public class AgentInitializerImpl implements AgentInitializer {
 
-  private final AdHocToolsSchemaResolver toolsSchemaResolver;
+  private static final Logger LOGGER = LoggerFactory.getLogger(AgentInitializerImpl.class);
+
+  private final AgentToolsResolver toolsResolver;
   private final GatewayToolHandlerRegistry gatewayToolHandlers;
 
   public AgentInitializerImpl(
-      AdHocToolsSchemaResolver toolsSchemaResolver,
-      GatewayToolHandlerRegistry gatewayToolHandlers) {
-    this.toolsSchemaResolver = toolsSchemaResolver;
+      AgentToolsResolver toolsResolver, GatewayToolHandlerRegistry gatewayToolHandlers) {
+    this.toolsResolver = toolsResolver;
     this.gatewayToolHandlers = gatewayToolHandlers;
   }
 
   @Override
   public AgentInitializationResult initializeAgent(AgentExecutionContext executionContext) {
     AgentContext agentContext =
-        Optional.ofNullable(executionContext.initialAgentContext()).orElseGet(AgentContext::empty);
+        Optional.ofNullable(executionContext.initialAgentContext())
+            .orElseGet(
+                () -> AgentContext.empty().withMetadata(executionContext.jobContext().metadata()));
 
     List<ToolCallResult> toolCallResults =
         Optional.ofNullable(executionContext.initialToolCallResults()).orElseGet(List::of);
 
     return switch (agentContext.state()) {
       case INITIALIZING -> initiateToolDiscovery(executionContext, agentContext, toolCallResults);
       case TOOL_DISCOVERY -> handleToolDiscoveryResults(agentContext, toolCallResults);
-      default -> new AgentContextInitializationResult(agentContext, toolCallResults);
+      default -> handleReadyState(executionContext, agentContext, toolCallResults);
     };
   }
 
-  private AgentInitializationResult initiateToolDiscovery(
+  private AgentInitializationResult handleReadyState(
       AgentExecutionContext executionContext,
       AgentContext agentContext,
       List<ToolCallResult> toolCallResults) {
-    final var toolElements = executionContext.toolElements();
 
-    // skip tool discovery if no tool elements are defined
-    if (CollectionUtils.isEmpty(toolElements)) {
-      return new AgentContextInitializationResult(
-          agentContext.withState(AgentState.READY), toolCallResults);
+    final var agentMetadata = agentContext.metadata();
+    final var executionMetadata = executionContext.jobContext().metadata();
+    if (agentMetadata == null
+        || !executionMetadata.processDefinitionKey().equals(agentMetadata.processDefinitionKey())) {
```

Comment:
This is basically the check if a migration has been applied. This will trigger for all running process instances (also without migration) as the data is missing, but as the resolving logic is cheap (using the variable on the sub-process and relying on cached data for the task) this should not be an issue.